coderd/authzquery/authz.go

```diff
@@ -0,0 +1,314 @@
+		return empty, execFunc(ctx, arg)
+	})
+	return func(ctx context.Context, arg ArgumentType) error {
+		_, err := f(ctx, arg)
+		return err
+	}
+}
+
+func authorizedFetchAndQuery[ObjectType rbac.Objecter, ArgumentType any,
+	Fetch func(ctx context.Context, arg ArgumentType) (ObjectType, error),
+	Query func(ctx context.Context, arg ArgumentType) (ObjectType, error)](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
+	fetchFunc Fetch,
+	queryFunc Query) Query {
+	return func(ctx context.Context, arg ArgumentType) (empty ObjectType, err error) {
+		// Fetch the rbac subject
+		act, ok := ActorFromContext(ctx)
+		if !ok {
+			return empty, NoActorError
+		}
+
+		// Fetch the database object
+		object, err := fetchFunc(ctx, arg)
+		if err != nil {
+			return empty, xerrors.Errorf("fetch object: %w", err)
+		}
+
+		// Authorize the action
+		err = authorizer.Authorize(ctx, act, action, object.RBACObject())
+		if err != nil {
+			return empty, LogNotAuthorizedError(ctx, logger, err)
+		}
+
+		return queryFunc(ctx, arg)
+	}
+}
+
+func authorizedFetch[ObjectType rbac.Objecter, ArgumentType any,
+	Fetch func(ctx context.Context, arg ArgumentType) (ObjectType, error)](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	fetchFunc Fetch) Fetch {
+	return authorizedQuery(logger, authorizer, rbac.ActionRead, fetchFunc)
+}
+
+// authorizedQuery is a generic function that wraps a database
+// query function (returns an object and an error) with authorization. The
+// returned function has the same arguments as the database function.
+//
+// The database query function will **ALWAYS** hit the database, even if the
+// user cannot read the resource. This is because the resource details are
+// required to run a proper authorization check.
+//
+// An optimized version of this could be written if the object's authz
+// subject properties are known by the caller.
+func authorizedQuery[ArgumentType any, ObjectType rbac.Objecter,
+	DatabaseFunc func(ctx context.Context, arg ArgumentType) (ObjectType, error)](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
+	f DatabaseFunc) DatabaseFunc {
+	return func(ctx context.Context, arg ArgumentType) (empty ObjectType, err error) {
+		// Fetch the rbac subject
+		act, ok := ActorFromContext(ctx)
+		if !ok {
+			return empty, NoActorError
+		}
+
+		// Fetch the database object
+		object, err := f(ctx, arg)
+		if err != nil {
+			return empty, xerrors.Errorf("fetch object: %w", err)
+		}
+
+		// Authorize the action
+		err = authorizer.Authorize(ctx, act, action, object.RBACObject())
+		if err != nil {
+			return empty, LogNotAuthorizedError(ctx, logger, err)
+		}
+
+		return object, nil
+	}
+}
+
+// authorizedFetchSet is like authorizedFetch, but works with lists of objects.
+// SQL filters are much more optimal.
+func authorizedFetchSet[ArgumentType any, ObjectType rbac.Objecter,
```

Comment:
```suggestion
func fetchWithPostFilter[ArgumentType any, ObjectType rbac.Objecter,
```

Reply:
I was confused if this was set like "set theory" or like "setting".  It's neither!